Extend "Whats new" modal for global nav

roo2 opened the pull request:
This PR is currently to demonstrate how the what's new modal can be extended to show the new content tour as described in: pfsHM7-cD-p2

This solves the technical challenge of loading the what's new modal without displaying the help center. Previously, it could only be shown if the help center slide out menu is also visible.

It also restyles the popup to be aligned to the bottom right and not interupt using calypso as much.

<img width="1728" alt="Screenshot 2024-02-26 at 5 04 28 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/30eca058-4822-46c3-8b01-e3c483b8d22f">

I'm adding logic that will keep track of which announcements the user has seen at a new endpoint `/whats-new/seen-announcement-ids` in D139800-code.

### Testing instructions
Apply D139800-code to your sandbox, and this PR.
Filter the network tab to `whats-new`
You should be able to see the modal initially, close it, and reopen it from "What's New" in the help menu

<img width="416" alt="Screenshot 2024-02-26 at 5 07 22 pm" src="https://github.com/Automattic/wp-calypso/assets/22446385/32750506-29d9-44d4-ad89-9472d8dfd982">

The whats new tour is also available within the editor via the ETK, this code had to be modified as well.

### To do in a follow up PR.
* Probably clean out old code relating to `HasSeenWhatsNewModal` which isn't useful. I see there is code for an indicator dot, I should be able to get this working as it was originally intended using the new `latestSeenWhatsNewModalItem` counter.
* Potentially change from using the older redux store pattern to a mutation?
* Confirm that the ETK whats-new javascript package is unused.

roo2 committed:
WIP: whats new modal for global nav

roo2 committed:
show critical notifications to users who have not yet seen them

roo2 commented:
> Looks good! What's needed for this part?
>> it will still need to be hooked up to only show automatically for users who haven't seen the critical announcements.


I worked on that today. I'm adding a "Critical" flag to the "whats new" announcements that come back from the API in D139640-code

I'm adding logic that will keep track of the highest `announcementId` that the user has seen at a new endpoint `/whats-new/latest-seen-item-id` D139800-code. 

If the user has not already seen an active `critical` announcement (or a more recent announcement) then the announcements dialog will be be opened by default.

eoigal commented:
All works for me so far. Are you going to include those todo's in this PR or in follow up PRs?

roo2 committed:
fix whats new box in the editor (ETK). Pass site_id to what-new announcements endpoint

roo2 commented:
I got the Editing toolkit issue fixed and added the filtering on `announcement.showTo === 'CLASSIC_NAV'` to D139800-code

Now this is _hopefully_ ready to test and ship and should do everything we need to support the new whats new tour.

I can clean up code around the old `hasSeenWhatsNewModal` flag in another PR. It looks like at some point it was showing a notification indicator dot when the user hadn't seen updates. I should be able to take this flag out completely and replace it with the new `latestSeenWhatsNewModalItem` counter.

Addison-Stavlo commented on apps/editing-toolkit/editing-toolkit-plugin/help-center/src/help-center.js:
```diff
@@ -81,6 +86,7 @@ function HelpCenterContent() {
 				</>
 			) }
 			<HelpCenter handleClose={ closeCallback } />
+			<WhatsNewGuide onClose={ handleCloseWhatsNew } />
```
Why are we adding this to the help center as opposed to somewhere else? (Calypso proper, adjusting the whats-new etk plugin, etc.).  

In general, the help center already adds and controls an instance of the `WhatNewGuide` at [a deeper level](https://github.com/Automattic/wp-calypso/blob/ea8ff3cc00e4cf6eea076d7afea7e71ff898bb35/packages/help-center/src/components/help-center-more-resources.tsx#L154). It doesn't seem like it should manage two separate instances of this guide at two different levels. And it sounds like our goal here is to make opening/managing this independent of the help center being open, this signals to me this instance of the WhatsNewGuide shouldn't be in the help center in the first place.

With whats-new as a package, [a separate ETK plugin](https://github.com/Automattic/wp-calypso/blob/ea8ff3cc00e4cf6eea076d7afea7e71ff898bb35/apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js#L43) as well, and our desire to be independent of help center, it seems like we could and should find this logic a better home and corresponding data store?

roo2 replied:
Good question! that's something I spend a lot of time working out.

I'm actually adding it again here because I'm taking it out of the [help-center package](https://github.com/Automattic/wp-calypso/pull/87875/files#diff-b3c3c91e9f478e4470a4a2106b69ddbfa8028619b1331c6b195361419de1c6cbL154)!

I moved it to by [loaded in calypso by the main layout](https://github.com/Automattic/wp-calypso/pull/87875/files#diff-7d1eb73ef67c978553b9d47cf17a8f4b9d93bd396ee6cb5ce89e75aedb8b0f19R130). And use the redux action setShowWhatsNewModal() to open it either from the [main layout on load](https://github.com/Automattic/wp-calypso/pull/87875/files#diff-7d1eb73ef67c978553b9d47cf17a8f4b9d93bd396ee6cb5ce89e75aedb8b0f19R117) or from the help-center package.

The problem is that the code running in the ETK does not share the redux store, so I need to ensure an instance of the whats-new modal is also loaded in the ETK's frame. 

So I need to load it in the ETK code, but not in a package that is shared with calypso, or else it would be loaded twice.

I also first thought to put it in the `whats-new` ETK package, but I found that this package doesn't actually seem to be active in the editor currently. It appears that the intent of the `whats-new` package is to add another menu item to the tools menu. It[ registers a javascript plugin called 'whats-new' ](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/whats-new/src/whats-new.js#L43C1-L43C15) but the plugin is currently never loaded.

It is also good that since the `whats-new` modal is used by the etk help center package, this package is responsible for loading it. So that way the package loads its own dependencies.

After this PR, I will look into deleting/migrating that whats-new ETK package because it seems to be unused, I asked in the Vertex slack if they have any more context on it. p1709169478558419-slack-C02T4NVL4JJ

Addison-Stavlo replied:
Ohh, that makes a bit more sense. Thanks for explaining 😄

Addison-Stavlo reviewed (commented):
Mostly the code seems reasonable but Im confused with the placement and conflation with help center directly.

Aside from that, Im a bit confused and concerned by some of the premise here, as I described more deeply on D139800-code

roo2 committed:
use array of announcement IDs instead of only saving the latest ID

roo2 commented:
Actually, it looks like the whats-new plugin is still used for some people, so I'll need to test and tidy up that package as well p1709169478558419-slack-C02T4NVL4JJ

Addison-Stavlo commented:
Testing this, the only whats-new guide that Pops up is the newsletter guide. But there are other active guides, and some marked as critical. Even after resetting the meta as mentioned on the diff, the only whats new guide I see is for the newsletter. it seems the items marked as active critical announcements are not making it through even after I have viewed and closed the newsletter one?

<img width="896" alt="Screenshot 2024-02-29 at 12 14 46 PM" src="https://github.com/Automattic/wp-calypso/assets/28742426/860e0728-97db-4376-a4b6-5d8c02dec94c">

Addison-Stavlo reviewed (commented):
One thing Im seeing in testing currently is nested unseen critical announcements seem to not surface.

* Reset user meta and cache
* load calypso
* see the "whats new" dialog pop up. Newsletter is on the front page, hit the "X" before viewing any other pages.
* reload the page. No whats new pops up.

Since the second page is a critical announcement (Jetpack AI set to critical atm), and hasn't been seen, should the whats-new open this announcement on load? 🤔 It seems this may have been marked as 'seen' even though it hasn't been.

After this I test the contextual announcement:
* Switch to a site with classic view enabled
* see whats-new pop up automatically again (Great! because the classic user contextual announcement had never been seen).
* unfortunately, it is showing the newsletter announcement and other announcements I have already seen first, and the first unseen announcement (a critical+contextual one that is causing this to open) is buried on page 3.

<hr/>

I wonder if we want to both:
* Ensure an announcement is not marked as seen if whats-new has not gotten to that page (and either had "next" or "x" acknowledging it).
* Filter the order of announcements, so seen announcements are listed after unseen announcements? Perhaps we want to list unseen critical announcements before unseen non-critical announcements as well?

Addison-Stavlo reviewed (commented):
Another issue to note:  This seems to break whats-new from the help center outside of calypso.

* Apply this build of ETK to your atomic site (can get this from the "Build Calypso Apps" check below -> details -> artifacts tab for the .zip file.
* Navigate to Wp-admin
* Open the help center
* click "whats-new"
* nothing opens. Before applying this ETK build, whats new does open on that side.

Addison-Stavlo commented:
I think my[ latter comment ](https://github.com/Automattic/wp-calypso/pull/87875#pullrequestreview-1909527686)may be the first thing we need to resolve here.

The [other issues](https://github.com/Automattic/wp-calypso/pull/87875#pullrequestreview-1909467612) I noted I think we definitely want to fix before we leave here to have a good whats new system that supports critical and contextual announcements, but even in this state this already seems to be a huge help for getting this to auto-open to surface new changes on global nav release.

roo2 committed:
move whats new site ID outside of the query to support moving it to the whats-new ETK package

roo2 commented:
I managed to get the whats-new ETK package working. 

Currently it's working in all places I could test including the ETK issue that you mentioned in your [other comment ](https://github.com/Automattic/wp-calypso/pull/87875#pullrequestreview-1909527686).

* Calypso show on load
* Calypso via the help menu
* ETK via the tools menu
* ETK via the help menu
* ETK (in calypso frame) show on load

But it's not ideal, currently we need to make sure it's loaded exactly once in the ETK and exactly once in calypso, so there is an implicict dependency between the `whats-new.js` and `help-center.js` ETK plugins that only one of them would load it. I've made this the `whats-new.js` plugin for now.

I had imagined this working like a singleton component, so only one instance of it would ever be loaded no matter how many times it was included. I just couldn't figure out how to do that or find any examples of us doing that. I think the original approach of having multiple instances of the modal and using flags to conditionally render them is the way to go.

I'll take a look this afternoon at refactoring it back to be conditionally rendered (take out the `setShowWhatsNewModal()` redux state)

roo2 commented:
>  wonder if we want to both:
> * Ensure an announcement is not marked as seen if whats-new has not gotten to that page (and either had "next" or "x" acknowledging it).
> * Filter the order of announcements, so seen announcements are listed after unseen announcements? Perhaps we want to list unseen critical announcements before unseen non-critical announcements as well?

I would hesitate to make the first change, because if there were two announcements, and the user always just dismisses the modal, the second announcement would never be marked as seen. So it would seem to them that the modal just keeps showing up without a "Do not show again" option.

I think the second change (showing critical announcements first) makes sense and would be a good improvement.

roo2 commented:
I also found a height issue on mobile which I will fix this afternoon

roo2 committed:
stop trying with singleton nonsense :)

roo2 commented:
I refactored so that everywhere that shows the modal will include its own version of the modal so there are no wierd cross/package dependencies. I also fixed the mobile height issue and retested everythign.

I think we could address the other issues you raised in a follow up PR :)

Addison-Stavlo committed:
Merge branch 'trunk' into try/whats-new-modal-for-global-nav

Addison-Stavlo commented on packages/whats-new/src/style.scss:
```diff
@@ -13,27 +13,31 @@ $wpcom-modal-footer-height: 30px + ( $wpcom-modal-content-spacing * 2 );
 			min-height: $wpcom-modal-content-min-height;
 			height: fit-content;
-			top: calc(50% - #{$wpcom-modal-footer-height * 0.5});
+			// top: calc(50% - #{$wpcom-modal-footer-height * 0.5});
```
^^

Addison-Stavlo commented on packages/whats-new/src/style.scss:
```diff
@@ -13,27 +13,31 @@ $wpcom-modal-footer-height: 30px + ( $wpcom-modal-content-spacing * 2 );
 
 	&.components-modal__frame {
-		top: calc(17.5vh - #{$wpcom-modal-footer-height * 0.5});
+		// top: calc(17.5vh - #{$wpcom-modal-footer-height * 0.5});
```
^^

Addison-Stavlo reviewed (approved):
I resolved the merge conflict so i could steal the ETK build from teamcity. 😁 

Testing:
* Whats new continues to work outside of calypso (ex - help center in wp-admin) now.
* Whats new appears on load in calypso when an announcement is unseen.

> I think we could address the other issues you raised in a follow up PR :)

Sounds good.

roo2 committed:
fix styles

roo2 committed:
only show critical announcements when new nav is enabled

roo2 committed:
dont show on checkout

roo2 committed:
remove false comments
